File: kit/cache-api/src/main/java/org/toolkit/spring/boot/cache/api/Cache.java
/* (C)2023*/
package org.toolkit.spring.boot.cache.api;

import java.time.Duration;
import java.util.Collection;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.TimeUnit;
import java.util.function.BiFunction;
import java.util.function.Function;
import lombok.NonNull;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public interface Cache<K, V> extends Iterable<Cache<K, V>> {

	void put(@NotNull @NonNull K k, V v);

	void put(@NotNull @NonNull K k, V v, Duration duration);

	void put(@NotNull @NonNull K k, V v, long delay, TimeUnit timeUnit);

	void putAll(@NotNull @NonNull Map<? extends K, ? extends V> data);

	int size();

	boolean isEmpty();

	boolean containsKey(@NotNull @NonNull K k);

	boolean putIfAbsent(@NotNull @NonNull K k, V v);

	void remove(@NotNull @NonNull K k);

	void removeAll(Collection<K> keys);

	void clear();

	@NotNull @NonNull Collection<K> keys();

	@NotNull @NonNull Collection<V> values();

	@Nullable V compute(
			@NotNull @NonNull K key, @NotNull @NonNull BiFunction<? super K, ? super V, ? extends V> remappingFunction);

	default @NotNull @NonNull Optional<V> computeOptional(
			@NotNull @NonNull K key,
			@NotNull @NonNull BiFunction<? super K, ? super V, ? extends V> remappingFunction) {
		return Optional.ofNullable(compute(key, remappingFunction));
	}

	@Nullable V computeIfAbsent(@NotNull @NonNull K key, @NotNull @NonNull Function<? super K, ? extends V> mappingFunction);

	default @NotNull @NonNull Optional<V> computeIfAbsentOptional(
			@NotNull @NonNull K key, @NotNull @NonNull Function<? super K, ? extends V> mappingFunction) {
		return Optional.ofNullable(computeIfAbsent(key, mappingFunction));
	}

	@Nullable V computeIfPresent(@NotNull @NonNull K key, Function<? super K, ? extends V> mappingFunction);

	default @NonNull Optional<V> computeIfPresentOptional(
			@NotNull @NonNull K key, @NotNull @NonNull Function<? super K, ? extends V> mappingFunction) {
		return Optional.ofNullable(computeIfPresent(key, mappingFunction));
	}

	boolean replace(@NotNull @NonNull K k, V v);

	@Nullable V get(@NotNull @NonNull K k);

	@NonNull @NotNull Map<? extends K, ? extends V> getAll();

	default Optional<V> getOptional(@NotNull @NonNull K k) {
		return Optional.ofNullable(get(k));
	}

	@Nullable V getAndRemove(@NotNull @NonNull K k);

	default Optional<V> getAndRemoveOptional(@NotNull @NonNull K k) {
		return Optional.ofNullable(getAndRemove(k));
	}

	default @NotNull @NonNull V getOrDefault(@NotNull @NonNull K k, V dv) {
		return Optional.ofNullable(get(k)).orElse(dv);
	}

	@NonNull @NotNull String getCacheName();
}

File: kit/cache-api/src/main/kotlin/org/toolkit4j/cache/Cache.kt
package org.toolkit4j.cache

import java.time.Duration
import java.util.*
import java.util.concurrent.TimeUnit
import java.util.function.BiFunction
import java.util.function.Function
import java.util.stream.Stream

interface Cache<K, V : Any?> : MutableIterable<Map.Entry<K, V>> {
    fun put(k: K, v: V?)

    fun put(k: K, v: V?, duration: Duration)

    fun put(k: K, v: V?, delay: Long, timeUnit: TimeUnit)

    fun putAll(data: Map<out K, V?>)

    fun putAll(data: Map<out K, V?>, duration: Duration)

    fun putAll(data: Map<out K, V?>, delay: Long, timeUnit: TimeUnit)

    fun size(): Int

    fun isEmpty(): Boolean

    fun containsKey(k: K): Boolean

    fun putIfAbsent(k: K, v: V?): Boolean

    fun remove(k: K)

    fun removeAll(keys: Collection<K>)

    fun clear()

    fun keys(): Collection<K>

    fun values(): Collection<V>

    fun compute(
        key: K, remappingFunction: BiFunction<in K, in V, out V>
    ): V?

    fun computeOptional(
        key: K,
        remappingFunction: BiFunction<in K, in V, out V>
    ): Optional<V & Any> {
        return Optional.ofNullable(compute(key, remappingFunction))
    }

    fun computeIfAbsent(key: K, mappingFunction: Function<in K, out V>): V?

    fun computeIfAbsentOptional(
        key: K, mappingFunction: Function<in K, out V>
    ): Optional<V & Any> {
        return Optional.ofNullable(computeIfAbsent(key, mappingFunction))
    }

    fun computeIfPresent(key: K, mappingFunction: Function<in K, out V>?): V?

    fun computeIfPresentOptional(
        key: K, mappingFunction: Function<in K, out V>
    ): Optional<V & Any> {
        return Optional.ofNullable(computeIfPresent(key, mappingFunction))
    }

    fun replace(k: K, v: V): Boolean

    fun get(k: K): V?

    fun getAll(): Map<out K, V>

    fun getOptional(k: K): Optional<V & Any> {
        return Optional.ofNullable(get(k))
    }

    fun getAndRemove(k: K): V?

    fun getAndRemoveOptional(k: K): Optional<V & Any> {
        return Optional.ofNullable(getAndRemove(k))
    }

    fun getOrDefault(k: K, v: V): V {
        return Optional.ofNullable(get(k)).orElse(v)
    }

    fun getCacheName(): String

    fun stream(): Stream<Map.Entry<K, V>>

    fun registerListener(listener: CacheListener<K, V>)

    fun unregisterListener(listener: CacheListener<K, V>)
}